File: sdks/python/src/lilypad/_utils/context_propagation.py
"""Context propagation utilities for distributed tracing.

This module provides utilities for extracting and injecting trace context
across process boundaries using OpenTelemetry propagators.
"""

from __future__ import annotations

import os
from typing import Any, MutableMapping

from opentelemetry import context, propagate
from opentelemetry.propagators.b3 import B3MultiFormat, B3SingleFormat
from opentelemetry.propagators.composite import CompositePropagator
from opentelemetry.propagators.jaeger import JaegerPropagator
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator


def get_propagator() -> CompositePropagator:
    """Get the configured propagator based on environment settings.

    Supports multiple propagation formats:
    - tracecontext (W3C Trace Context) - default
    - b3 (B3 Single Format)
    - b3multi (B3 Multi Format)
    - jaeger (Jaeger)
    - composite (all of the above)

    Returns:
        A configured propagator instance
    """
    propagator_type = os.getenv("LILYPAD_PROPAGATOR", "tracecontext").lower()

    if propagator_type == "b3":
        return CompositePropagator([B3SingleFormat()])
    elif propagator_type == "b3multi":
        return CompositePropagator([B3MultiFormat()])
    elif propagator_type == "jaeger":
        return CompositePropagator([JaegerPropagator()])
    elif propagator_type == "composite":
        # Support all common formats
        return CompositePropagator(
            [
                TraceContextTextMapPropagator(),
                B3SingleFormat(),
                B3MultiFormat(),
                JaegerPropagator(),
            ]
        )
    else:
        # Default to W3C Trace Context
        return CompositePropagator([TraceContextTextMapPropagator()])


class ContextPropagator:
    """Utility class for extracting and injecting trace context.

    WARNING: This class modifies global OpenTelemetry state by calling
    `propagate.set_global_textmap()`. This affects all OpenTelemetry
    instrumentation in the current process. If you have other OpenTelemetry
    integrations, they will use the propagator configured here.

    The propagator type can be controlled via the LILYPAD_PROPAGATOR
    environment variable. Supported values:
    - "tracecontext" (default): W3C Trace Context
    - "b3": B3 Single Format
    - "b3multi": B3 Multi Format
    - "jaeger": Jaeger format
    - "composite": All formats (for maximum compatibility)
    """

    def __init__(self, set_global: bool = True) -> None:
        """Initialize the context propagator.

        Args:
            set_global: Whether to set the global OpenTelemetry propagator.
                       Default is True for backward compatibility.

        WARNING: If set_global is True, this sets the global OpenTelemetry
        propagator, affecting all trace context propagation in the current process.
        """
        self.propagator = get_propagator()
        # Set the global propagator - this affects the entire process
        if set_global:
            propagate.set_global_textmap(self.propagator)

    def extract_context(self, carrier: dict[str, Any]) -> context.Context:
        """Extract trace context from a carrier (e.g., HTTP headers).

        Args:
            carrier: A dictionary-like object containing trace context
                    (typically HTTP headers)

        Returns:
            The extracted context
        """
        return propagate.extract(carrier)

    def inject_context(self, carrier: MutableMapping[str, str], context: context.Context | None = None) -> None:
        """Inject current trace context into a carrier.

        Args:
            carrier: A mutable mapping to inject context into
                    (typically HTTP headers)
            context: Optional context to inject. If None, uses current context
        """
        propagate.inject(carrier, context=context)

    def with_extracted_context(self, carrier: dict[str, Any]) -> tuple[context.Context, object | None]:
        """Extract context and attach it, returning the context and token.

        This is useful for manual context management.

        Args:
            carrier: A dictionary-like object containing trace context

        Returns:
            Tuple of (extracted context, detach token)
        """
        ctx = self.extract_context(carrier)
        token = context.attach(ctx)
        return ctx, token

    def detach_context(self, token: object | None) -> None:
        """Detach a previously attached context.

        Args:
            token: The token returned from attach operation
        """
        if token is not None:
            context.detach(token)


# Global instance for convenience - created lazily
_propagator: ContextPropagator | None = None


def _get_propagator() -> ContextPropagator:
    """Get or create the global propagator instance."""
    global _propagator
    if _propagator is None:
        # Check if we should set global propagator
        # This is controlled by whether configure() has been called
        import os

        set_global = os.environ.get("_LILYPAD_PROPAGATOR_SET_GLOBAL", "true").lower() == "true"
        _propagator = ContextPropagator(set_global=set_global)
    return _propagator


def _extract_context(carrier: dict[str, Any]) -> context.Context:
    """Extract trace context from a carrier (e.g., HTTP headers).

    This function extracts OpenTelemetry trace context from a carrier object,
    typically HTTP headers received from a request. This allows distributed
    traces to maintain parent-child relationships across service boundaries.

    Args:
        carrier: A dictionary-like object containing trace context headers.
                 Typically HTTP headers from an incoming request.

    Returns:
        An OpenTelemetry Context object containing the extracted trace information.
        This context can be used to create child spans or passed to other functions.

    Example:
        Basic usage in a web handler:

        ```python
        from fastapi import FastAPI, Request
        from lilypad import trace
        from lilypad._utils.context_propagation import extract_context

        app = FastAPI()


        @app.post("/process")
        async def process_request(request: Request, data: dict):
            # Extract trace context from incoming request headers
            context = extract_context(dict(request.headers))

            # Use the extracted context in a traced function
            @trace(parent_context=context)
            async def process_data(data: dict) -> dict:
                # This span will be a child of the incoming trace
                result = await do_processing(data)
                return result

            return await process_data(data)
        ```

        Manual context management:

        ```python
        from opentelemetry import context as otel_context
        from lilypad._utils.context_propagation import extract_context


        def handle_message(headers: dict, payload: dict):
            # Extract context from message headers
            ctx = extract_context(headers)

            # Attach the context for this scope
            token = otel_context.attach(ctx)
            try:
                # All traces within this block will use the extracted context
                process_message(payload)
            finally:
                # Detach context when done
                otel_context.detach(token)
        ```

        Using with Flask:

        ```python
        from flask import Flask, request
        from lilypad import trace
        from lilypad._utils.context_propagation import extract_context

        app = Flask(__name__)


        @app.route("/api/endpoint", methods=["POST"])
        def api_endpoint():
            # Extract context from Flask request headers
            context = extract_context(dict(request.headers))

            @trace(parent_context=context)
            def handle_request(data):
                # Process as child of incoming trace
                return {"status": "processed", "data": data}

            return handle_request(request.json)
        ```

    Note:
        The function looks for trace headers based on the configured propagator:
        - W3C TraceContext: looks for 'traceparent' header
        - B3: looks for 'b3' or 'x-b3-*' headers
        - Jaeger: looks for 'uber-trace-id' header

        If no valid trace context is found, returns an empty context.
    """
    return _get_propagator().extract_context(carrier)


def _inject_context(carrier: MutableMapping[str, str], context: context.Context | None = None) -> None:
    """Inject current trace context into a carrier (e.g., HTTP headers).

    This function injects the current OpenTelemetry trace context into a carrier
    object, typically HTTP headers. This allows trace context to be propagated
    across service boundaries in distributed systems.

    Args:
        carrier: A mutable mapping (e.g., dict) to inject trace context into.
                 Typically HTTP headers that will be sent with a request.
        context: Optional specific context to inject. If None (default),
                 uses the current active context.

    Returns:
        None. The carrier is modified in-place with trace headers.

    Example:
        Basic usage with HTTP requests:

        ```python
        import httpx
        from lilypad import trace
        from lilypad._utils.context_propagation import inject_context


        @trace()
        async def call_external_service(data: dict) -> dict:
            # Create headers dict
            headers = {}

            # Inject current trace context into headers
            inject_context(headers)
            # Now headers contains: {'traceparent': '00-trace_id-span_id-01'}

            # Make HTTP request with trace context
            async with httpx.AsyncClient() as client:
                response = await client.post("https://api.example.com/process", json=data, headers=headers)
                return response.json()
        ```

        Using with requests library:

        ```python
        import requests
        from lilypad import trace
        from lilypad._utils.context_propagation import inject_context


        @trace()
        def sync_call_service(payload: dict) -> dict:
            headers = {"Content-Type": "application/json"}

            # Add trace context to existing headers
            inject_context(headers)

            response = requests.post("https://api.example.com/endpoint", json=payload, headers=headers)
            return response.json()
        ```

        Injecting specific context:

        ```python
        from opentelemetry import context as otel_context

        # Capture context from one trace
        saved_context = otel_context.get_current()

        # Later, inject that specific context
        headers = {}
        inject_context(headers, context=saved_context)
        ```

    Note:
        The exact headers added depend on the configured propagator format:
        - W3C TraceContext: adds 'traceparent' (and optionally 'tracestate')
        - B3: adds 'b3' header (single format) or 'x-b3-*' headers (multi format)
        - Jaeger: adds 'uber-trace-id' header
    """
    _get_propagator().inject_context(carrier, context)


def with_extracted_context(carrier: dict[str, Any]) -> tuple[context.Context, object | None]:
    """Extract context and attach it, returning the context and token.

    Args:
        carrier: A dictionary-like object containing trace context

    Returns:
        Tuple of (extracted context, detach token)
    """
    return _get_propagator().with_extracted_context(carrier)


def detach_context(token: object | None) -> None:
    """Detach a previously attached context.

    Args:
        token: The token returned from attach operation
    """
    _get_propagator().detach_context(token)

File: sdks/typescript/src/utils/context-propagation.ts
/**
 * Context propagation utilities for distributed tracing.
 *
 * This module provides utilities for extracting and injecting trace context
 * across process boundaries using OpenTelemetry propagators.
 */

import { Context, propagation, TextMapGetter, TextMapSetter, context } from '@opentelemetry/api';
import { CompositePropagator, W3CTraceContextPropagator } from '@opentelemetry/core';
import { B3Propagator, B3InjectEncoding } from '@opentelemetry/propagator-b3';
import { JaegerPropagator } from '@opentelemetry/propagator-jaeger';
import { logger } from './logger';

import type { Propagator } from '../types';

/**
 * Get the configured propagator based on environment settings.
 *
 * Supports multiple propagation formats:
 * - tracecontext (W3C Trace Context) - default
 * - b3 (B3 Single Format)
 * - b3multi (B3 Multi Format)
 * - jaeger (Jaeger)
 * - composite (all of the above)
 *
 * @returns A configured propagator instance
 */
export function getPropagator(type?: Propagator): CompositePropagator {
  const propagatorType = (
    type ||
    process.env.LILYPAD_PROPAGATOR ||
    'tracecontext'
  ).toLowerCase() as Propagator;

  switch (propagatorType) {
    case 'b3':
      return new CompositePropagator({
        propagators: [new B3Propagator({ injectEncoding: B3InjectEncoding.SINGLE_HEADER })],
      });
    case 'b3multi':
      return new CompositePropagator({
        propagators: [new B3Propagator({ injectEncoding: B3InjectEncoding.MULTI_HEADER })],
      });
    case 'jaeger':
      return new CompositePropagator({
        propagators: [new JaegerPropagator()],
      });
    case 'composite':
      // Support all common formats
      return new CompositePropagator({
        propagators: [
          new W3CTraceContextPropagator(),
          new B3Propagator({ injectEncoding: B3InjectEncoding.SINGLE_HEADER }),
          new B3Propagator({ injectEncoding: B3InjectEncoding.MULTI_HEADER }),
          new JaegerPropagator(),
        ],
      });
    default:
      // Default to W3C Trace Context
      return new CompositePropagator({
        propagators: [new W3CTraceContextPropagator()],
      });
  }
}

/**
 * Utility class for extracting and injecting trace context.
 *
 * WARNING: This class modifies global OpenTelemetry state by calling
 * `propagation.setGlobalPropagator()`. This affects all OpenTelemetry
 * instrumentation in the current process. If you have other OpenTelemetry
 * integrations, they will use the propagator configured here.
 *
 * The propagator type can be controlled via the LILYPAD_PROPAGATOR
 * environment variable. Supported values:
 * - "tracecontext" (default): W3C Trace Context
 * - "b3": B3 Single Format
 * - "b3multi": B3 Multi Format
 * - "jaeger": Jaeger format
 * - "composite": All formats (for maximum compatibility)
 */
export class ContextPropagator {
  private propagator: CompositePropagator;

  constructor(setGlobal: boolean = true, propagatorType?: Propagator) {
    this.propagator = getPropagator(propagatorType);

    // Set the global propagator - this affects the entire process
    if (setGlobal) {
      propagation.setGlobalPropagator(this.propagator);
      logger.debug(`Set global propagator to: ${propagatorType || 'tracecontext'}`);
    }
  }

  /**
   * Extract trace context from a carrier (e.g., HTTP headers).
   *
   * @param carrier - A dictionary-like object containing trace context
   *                 (typically HTTP headers)
   * @returns The extracted context
   */
  extractContext(carrier: Record<string, unknown>): Context {
    const getter: TextMapGetter = {
      keys: (carrier) => Object.keys(carrier),
      get: (carrier, key) => {
        const value = carrier[key];
        return Array.isArray(value) ? value[0] : value;
      },
    };

    return propagation.extract(context.active(), carrier, getter);
  }

  /**
   * Inject current trace context into a carrier.
   *
   * @param carrier - A mutable mapping to inject context into
   *                 (typically HTTP headers)
   * @param context - Optional context to inject. If undefined, uses current context
   */
  injectContext(carrier: Record<string, string>, contextParam?: Context): void {
    const setter: TextMapSetter = {
      set: (carrier, key, value) => {
        carrier[key] = value;
      },
    };

    propagation.inject(contextParam || context.active(), carrier, setter);
  }

  /**
   * Extract context and attach it, returning the context and token.
   *
   * This is useful for manual context management.
   *
   * @param carrier - A dictionary-like object containing trace context
   * @returns Tuple of [extracted context, detach token]
   */
  withExtractedContext(carrier: Record<string, unknown>): [Context, (() => void) | undefined] {
    const ctx = this.extractContext(carrier);

    // We'll return a function to restore the previous context
    const restoreFn = () => {
      // This is a workaround since JS API doesn't have attach/detach
      // In practice, you would use context.with() to run code in a context
    };

    return [ctx, restoreFn];
  }

  /**
   * Detach a previously attached context.
   *
   * @param token - The restore function returned from withExtractedContext
   */
  detachContext(token: (() => void) | undefined): void {
    if (token !== undefined) {
      token();
    }
  }
}

// Global instance for convenience - created lazily
let _propagator: ContextPropagator | null = null;

/**
 * Get or create the global propagator instance.
 */
function _getPropagator(): ContextPropagator {
  if (_propagator === null) {
    // Check if we should set global propagator
    // This is controlled by whether configure() has been called
    const setGlobal = process.env._LILYPAD_PROPAGATOR_SET_GLOBAL !== 'false';
    _propagator = new ContextPropagator(setGlobal);
  }
  return _propagator;
}

/**
 * Extract trace context from a carrier (e.g., HTTP headers).
 *
 * This function extracts OpenTelemetry trace context from a carrier object,
 * typically HTTP headers received from a request. This allows distributed
 * traces to maintain parent-child relationships across service boundaries.
 *
 * @param carrier - A dictionary-like object containing trace context headers.
 *                 Typically HTTP headers from an incoming request.
 * @returns An OpenTelemetry Context object containing the extracted trace information.
 *
 * @example
 * Basic usage in an Express handler:
 * ```ts
 * import express from 'express';
 * import { trace } from '@lilypad/sdk';
 * import { extractContext } from '@lilypad/sdk/utils/context-propagation';
 *
 * const app = express();
 *
 * app.post('/process', async (req, res) => {
 *   // Extract trace context from incoming request headers
 *   const context = extractContext(req.headers);
 *
 *   // Use the extracted context in a traced function
 *   const processData = trace(async (data: any) => {
 *     // This span will be a child of the incoming trace
 *     const result = await doProcessing(data);
 *     return result;
 *   });
 *
 *   const result = await processData(req.body);
 *   res.json(result);
 * });
 * ```
 */
export function extractContext(carrier: Record<string, unknown>): Context {
  return _getPropagator().extractContext(carrier);
}

/**
 * Inject current trace context into a carrier (e.g., HTTP headers).
 *
 * This function injects the current OpenTelemetry trace context into a carrier
 * object, typically HTTP headers. This allows trace context to be propagated
 * across service boundaries in distributed systems.
 *
 * @param carrier - A mutable mapping (e.g., object) to inject trace context into.
 *                 Typically HTTP headers that will be sent with a request.
 * @param context - Optional specific context to inject. If undefined,
 *                 uses the current active context.
 *
 * @example
 * Basic usage with fetch:
 * ```ts
 * import { trace } from '@lilypad/sdk';
 * import { injectContext } from '@lilypad/sdk/utils/context-propagation';
 *
 * const callExternalService = trace(async (data: any) => {
 *   // Create headers object
 *   const headers: Record<string, string> = {
 *     'Content-Type': 'application/json',
 *   };
 *
 *   // Inject current trace context into headers
 *   injectContext(headers);
 *   // Now headers contains: {'traceparent': '00-trace_id-span_id-01'}
 *
 *   // Make HTTP request with trace context
 *   const response = await fetch('https://api.example.com/process', {
 *     method: 'POST',
 *     headers,
 *     body: JSON.stringify(data),
 *   });
 *
 *   return response.json();
 * });
 * ```
 */
export function injectContext(carrier: Record<string, string>, context?: Context): void {
  _getPropagator().injectContext(carrier, context);
}

/**
 * Configure the global propagator for the SDK.
 * This should be called during SDK initialization.
 *
 * @param propagatorType - The type of propagator to use
 * @param setGlobal - Whether to set as global propagator (default: true)
 */
export function configurePropagator(propagatorType?: Propagator, setGlobal: boolean = true): void {
  if (setGlobal) {
    // Create new propagator and set it globally
    _propagator = new ContextPropagator(true, propagatorType);
  }
}
